File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/util/ConfigPrefHelper.java
package de.dimedis.mobileentry.util;

import android.text.TextUtils;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;

import java.util.List;
import java.util.Map;

import de.dimedis.mobileentry.BuildConfig;
import de.dimedis.mobileentry.ConfigPref;
import de.dimedis.mobileentry.MobileEntryApplication;
import de.dimedis.mobileentry.backend.response.Border;
import de.dimedis.mobileentry.backend.response.Localize;
import de.dimedis.mobileentry.backend.response.UserPrefs;
import de.dimedis.mobileentry.backend.response.Versions;

public class ConfigPrefHelper {

    static public Map<String, Localize> getLanguages() {
        Gson gson = new Gson();
        String container = MobileEntryApplication.getConfigPreferences().languagesContainer2();
        return gson.fromJson(container, new TypeToken<Map<String, Localize>>() {
        }.getType());
    }

    static public void setLanguages(Map<String, Localize> languages) {
        MobileEntryApplication.getConfigPreferences().setLanguagesContainer2(new Gson().toJson(languages));
    }

    static public Versions getVersions() {
        String versionsContainer = MobileEntryApplication.getConfigPreferences().versionsContainer();  //versionsFromServerContainer();
        int versionCode = BuildConfig.VERSION_CODE;
        String versionName = BuildConfig.VERSION_NAME;
        String libVersion = ConfigPrefHelper.getLibVersion();

        Versions defaultVersion = new Versions();

        if (TextUtils.isEmpty(versionsContainer)) {
            defaultVersion.setLanguages("0.01");
            defaultVersion.setMyAvailableBorders("2015090116:00:54.980");
            defaultVersion.setLocalConfig("0");
            defaultVersion.setSettings("0");
            MobileEntryApplication.getConfigPreferences().setVersionsContainer(new Gson().toJson(defaultVersion));  //(version));
        } else {
            defaultVersion = new Gson().fromJson(versionsContainer, Versions.class);
        }
        defaultVersion.setApp(String.format("%s (%d)", versionName, versionCode));
        defaultVersion.setLibrary(libVersion);
        return defaultVersion;
    }

    private static String getLibVersion() {
        return PrefUtils.getLibraryVersion();
    }

    static public Versions getVersionsFromServer() {
        String versionsFromServerContainer = MobileEntryApplication.getConfigPreferences().versionsFromServerContainer();
        if (TextUtils.isEmpty(versionsFromServerContainer)) {
            MobileEntryApplication.getConfigPreferences().setVersionsFromServerContainer(new Gson().toJson(getVersions()));
        }
        return new Gson().fromJson(versionsFromServerContainer, Versions.class);
    }
    static public boolean isUpdateAvailable() {
        return !(getVersions().getLibrary().equals(getVersionsFromServer().getLibrary()) && getVersions().getApp().equals(getVersionsFromServer().getApp()));
    }

    static public void setVersions(Versions version) {
        MobileEntryApplication.getConfigPreferences().setVersionsContainer(new Gson().toJson(version));
    }

    static public void setServerVersions(Versions version) {
        MobileEntryApplication.getConfigPreferences().setVersionsFromServerContainer(new Gson().toJson(version));
    }

    static public List<Border> getBorders() {
        return new Gson().fromJson(MobileEntryApplication.getConfigPreferences().bordersContainer(), new TypeToken<List<Border>>() {
        }.getType());
    }

    static public void setBorders(List<Border> borders) {
        MobileEntryApplication.getConfigPreferences().setBordersContainer(new Gson().toJson(borders));
    }

    static public ConfigPref getPref() {
        return MobileEntryApplication.getConfigPreferences();
    }

    static public Border getUsersBorder() {
        try {
            return new Gson().fromJson(MobileEntryApplication.getConfigPreferences().usersBorder(), Border.class);
        } catch (JsonSyntaxException jse) {
            Log.e("ConfigPrefHelper", "getUsersBorder()", jse);
        }
        return null;
    }

    static public void setUsersBorder(Border border) {
        MobileEntryApplication.getConfigPreferences().setUsersBorder(new Gson().toJson(border));
        PrefUtils.setBorderName(border.getBorderName());
    }

    static public void setUserFunctions(List<String> list) {
        MobileEntryApplication.getConfigPreferences().setListUserFunctions(new Gson().toJson(list));
    }

    static public List<String> getUserFunctions() {
        return new Gson().fromJson(MobileEntryApplication.getConfigPreferences().listUserFunctions(), new TypeToken<List<String>>() {
        }.getType());
    }

    static public void setUserPrefs(UserPrefs userPrefs) {
        String newPrefs = new Gson().toJson(userPrefs);
        MobileEntryApplication.getConfigPreferences().setUserPrefs(newPrefs);
        try {
            Logger.asyncLog("Saving new F-KEYS", String.format("New set of F-KEYS: %s;", newPrefs), LogType.INFO);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    static public UserPrefs getUserPrefs() {
        return new Gson().fromJson(MobileEntryApplication.getConfigPreferences().userPrefs(), UserPrefs.class);
    }

    static public void resetDevice() {
        PrefUtils.setInitCompleted(false);
        PrefUtils.setLoginCompleted(false);
    }

    static public void userLogout() {
        PrefUtils.setLoginCompleted(false);
    }
}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/util/ConfigPrefHelper.kt
package de.dimedis.mobileentry.util

import android.text.TextUtils
import android.util.Log
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import de.dimedis.mobileentry.BuildConfig
import de.dimedis.mobileentry.ConfigPref
import de.dimedis.mobileentry.backend.response.Border
import de.dimedis.mobileentry.backend.response.Localize
import de.dimedis.mobileentry.backend.response.UserPrefs
import de.dimedis.mobileentry.backend.response.Versions

object ConfigPrefHelper {
    fun getLanguages(): Map<String?, Localize?>? {
        val gson = Gson()
        val container: String? = ConfigPref.languagesContainer2
        return gson.fromJson(container, object : TypeToken<Map<String?, Localize?>?>() {}.type)
    }

    fun setLanguages(languages: HashMap<String?, Localize>?) {
        ConfigPref.languagesContainer2 = Gson().toJson(languages)
    }

    fun getVersions(): Versions {
        val versionsContainer: String? = ConfigPref.versionsContainer //versionsFromServerContainer();
        val versionCode = BuildConfig.VERSION_CODE
        val versionName = BuildConfig.VERSION_NAME
        val libVersion = getLibVersion()
        var defaultVersion = Versions()
        if (TextUtils.isEmpty(versionsContainer)) {
            defaultVersion.languages = "0.01"
            defaultVersion.myAvailableBorders = "2015090116:00:54.980"
            defaultVersion.localConfig = "0"
            defaultVersion.settings = "0"
            ConfigPref.versionsContainer = Gson().toJson(defaultVersion) //(version));
        } else {
            defaultVersion = Gson().fromJson(versionsContainer, Versions::class.java)
        }
        defaultVersion.app = String.format(versionName, versionCode)
        defaultVersion.library = libVersion
        return defaultVersion
    }

    private fun getLibVersion(): String? {
        return PrefUtils.getLibraryVersion() // libraryVersion
    }

    fun getVersionsFromServer(): Versions {
        val versionsFromServerContainer: String? = ConfigPref.versionsFromServerContainer
        if (TextUtils.isEmpty(versionsFromServerContainer)) {
            ConfigPref.versionsFromServerContainer = (Gson().toJson(getVersions()))
        }
        return Gson().fromJson(versionsFromServerContainer, Versions::class.java)
    }

    fun isUpdateAvailable(): Boolean {
        return !(getVersions().library == getVersionsFromServer().library && getVersions().app!! >= getVersionsFromServer().app!!)
    }

    fun setVersions(version: Versions?) {
        ConfigPref.versionsContainer //(Gson().toJson(version))
    }

    fun setServerVersions(version: Versions?) {
        /*   version.setLibrary("2.2.0");
        version.setLanguages("2018060114:06:47");
        version.setSettings("2018052823:07:68");
        version.setLocalConfig("2018052823:07:71");
        version.setMyAvailableBorders("2018052823:07:63");*/
        ConfigPref.versionsFromServerContainer = (Gson().toJson(version))
    }

    fun getBorders(): List<Border> {
        return Gson().fromJson(ConfigPref.bordersContainer, object : TypeToken<List<Border?>?>() {}.type)
    }

    fun setBorders(borders: List<Border?>?) {
        ConfigPref.bordersContainer = (Gson().toJson(borders))
    }

    fun getUsersBorder(): Border? {
        try {
            return Gson().fromJson(ConfigPref.usersBorder, Border::class.java)
        } catch (jse: JsonSyntaxException) {
            Log.e("ConfigPrefHelper", "getUsersBorder()", jse)
        }
        return null
    }

    fun setUsersBorder(border: Border) {
        ConfigPref.usersBorder = (Gson().toJson(border))
        PrefUtils.setBorderName(border.borderName)
    }

    fun setUserFunctions(list: List<String?>?) {
        ConfigPref.listUserFunctions = (Gson().toJson(list))
    }

    fun getUserFunctions(): List<String> {
        return Gson().fromJson(ConfigPref.listUserFunctions ?: "[]", object : TypeToken<List<String?>?>() {}.type)
    }

    fun setUserPrefs(userPrefs: UserPrefs?) {
        val newPrefs = Gson().toJson(userPrefs)
        ConfigPref.userPrefs = (newPrefs)
        try {
            Logger.asyncLog("Saving new F-KEYS", String.format("New set of F-KEYS: %s;", newPrefs), LogType.INFO)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    fun getUserPrefs(): UserPrefs {
        return Gson().fromJson(ConfigPref.userPrefs, UserPrefs::class.java)
    }

    fun resetDevice() {
        PrefUtils.setInitCompleted(false)
        PrefUtils.setLoginCompleted(false)
        PrefUtils.setKioskModeEnabled(false)
    }

    fun userLogout() {
        PrefUtils.setLoginCompleted(false)
    }
}